src/envs/coffee.py

```diff
@@ -0,0 +1,756 @@
+
+        # Options
+        self._PickJug = ParameterizedOption(
+            "PickJug",
+            types=[self._robot_type, self._jug_type],
+            params_space=Box(0, 1, (0, )),
+            policy=self._PickJug_policy,
+            initiable=lambda s, m, o, p: True,
+            terminal=self._PickJug_terminal,
+        )
+        self._PlaceJugInMachine = ParameterizedOption(
+            "PlaceJugInMachine",
+            types=[self._robot_type, self._jug_type, self._machine_type],
+            params_space=Box(0, 1, (0, )),
+            policy=self._PlaceJugInMachine_policy,
+            initiable=lambda s, m, o, p: True,
+            terminal=self._PlaceJugInMachine_terminal,
+        )
+        self._TurnMachineOn = ParameterizedOption(
+            "TurnMachineOn",
+            types=[self._robot_type, self._machine_type],
+            params_space=Box(0, 1, (0, )),
+            policy=self._TurnMachineOn_policy,
+            initiable=lambda s, m, o, p: True,
+            terminal=self._TurnMachineOn_terminal,
+        )
+        self._Pour = ParameterizedOption(
+            "Pour",
+            types=[self._robot_type, self._jug_type, self._cup_type],
+            params_space=Box(0, 1, (0, )),
+            policy=self._Pour_policy,
+            initiable=lambda s, m, o, p: True,
+            terminal=self._Pour_terminal,
+        )
+
+        # Static objects (always exist no matter the settings).
+        self._robot = Object("robby", self._robot_type)
+        self._jug = Object("juggy", self._jug_type)
+        self._machine = Object("coffee_machine", self._machine_type)
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "coffee"
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        next_state = state.copy()
+        norm_dx, norm_dy, norm_dz, norm_dtilt, norm_dfingers = action.arr
+        # Denormalize the action.
+        dx = norm_dx * self.max_position_vel
+        dy = norm_dy * self.max_position_vel
+        dz = norm_dz * self.max_position_vel
+        dtilt = norm_dtilt * self.max_angular_vel
+        dfingers = norm_dfingers * self.max_finger_vel
+        # Apply changes to the robot, taking bounds into account.
+        x = np.clip(state.get(self._robot, "x") + dx, self.x_lb, self.x_ub)
+        y = np.clip(state.get(self._robot, "y") + dy, self.y_lb, self.y_ub)
+        z = np.clip(state.get(self._robot, "z") + dz, self.z_lb, self.z_ub)
+        current_tilt = state.get(self._robot, "tilt")
+        tilt = np.clip(current_tilt + dtilt, self.tilt_lb, self.tilt_ub)
+        current_fingers = state.get(self._robot, "fingers")
+        fingers = np.clip(current_fingers + dfingers, self.closed_fingers,
+                          self.open_fingers)
+        # The deltas may be outdated because of the clipping, so recompute
+        # or delete them.
+        dx = x - state.get(self._robot, "x")
```

Comment:
we only need dx, dy here because we want to apply the same change to the jug if it is held? but then why don't we need dz too?